File: docs/algorithms/code/algorithms/getMinimumDifference.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function getMinimumDifference(root: TreeNode | null): number {
  const res: number[] = [];
  const stack: TreeNode[] = [];

  let current: TreeNode | null = root;

  while (current || stack.length) {
    while (current) {
      stack.push(current);
      current = current.left;
    }

    current = stack.pop()!;
    res.push(current.val);
    current = current.right;
  }

  let min = res[1] - res[0];

  for (let i = 2; i < res.length; i++) {
    min = Math.min(min, res[i] - res[i - 1]);
  }

  return min;
}

File: docs/algorithms/code/algorithms/getMinimumDifference.go
// 思路：先中序遍历，然后求最小值
// 复杂度：时间复杂度O(n)，空间复杂度O(n)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func getMinimumDifference(root *TreeNode) int {
	if root == nil {
		return 0
	}

	treeValue := []int{}

	stack := []*TreeNode{}
	for len(stack) > 0 || root != nil {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]
		treeValue = append(treeValue, root.Val)
		stack = stack[:len(stack)-1]
		root = root.Right
	}

	min := treeValue[1] - treeValue[0]
	for i := 2; i < len(treeValue); i++ {
		diff := treeValue[i] - treeValue[i-1]
		if diff < min  {
			min = diff
		}
	}

	return min
}